### PR TITLE
to_device: Handle nested lists/tuples recursively

### PR DESCRIPTION
**Pull request by ottonemo**
The previous implementation of `to_device` would break when a
user decided to return a list of tensors in `forward`.

This patch applies `to_device` recursively and adds support
for lists in addition to tuples.

### CODE REVIEW

File: skorch/tests/test_utils.py [hunk 1]
```diff
@@ -230,6 +234,32 @@ def test_check_device_packed_padded_sequence(
...
 
+    @pytest.mark.parametrize('device_from, device_to', [
+        ('cpu', 'cpu'),
+        ('cpu', 'cuda'),
+        ('cuda', 'cpu'),
+        ('cuda', 'cuda'),
+        (None, None),
+    ])
+    def test_nested_data(self, to_device, x_list, device_from, device_to):
+        # Sometimes data is nested because it would need to be padded so its
```
Comment (BenjaminBossan) [inline skorch/tests/test_utils.py:245]:
```suggestion
        # Sometimes data is nested because it would need to be padded so it's
```

**Comment by ottonemo** [comment]
> Just a heads up, there is also #657 working on `to_device`. Does it make sense to also adjust `to_numpy` with the same logic?

I think it makes sense for `predict_proba` and `transform`. I will add it.

> Also, it seems that when passing a list, `to_device` will return a tuple. Do you think it should be a list instead?

Fixed, thanks.

**Comment by ottonemo** [comment]
@BenjaminBossan please review again.

**Comment by ottonemo** [comment]
@BenjaminBossan I think there's an issue with this when calling `to_numpy([1,2,3])`. Before this returned a numpy array, now it doesn't anymore. Please don't merge this yet.

**Comment by BenjaminBossan** [comment]
> I think there's an issue with this when calling `to_numpy([1,2,3])`. Before this returned a numpy array, now it doesn't anymore. Please don't merge this yet.

Is it correct that this raises a `TypeError` (old and new implementation)? I wonder if that's the desired behavior and if it is, if we should clarify that with a corresponding test.

**Comment by ottonemo** [comment]
> > I think there's an issue with this when calling `to_numpy([1,2,3])`. Before this returned a numpy array, now it doesn't anymore. Please don't merge this yet.
> 
> Is it correct that this raises a `TypeError` (old and new implementation)? I wonder if that's the desired behavior and if it is, if we should clarify that with a corresponding test.

This is a question of scope. `to_numpy` was originally introduced to convert from PyTorch to numpy and included device handling / gradient detaching. Now we are transitioning away from that definition by handling other data types as well but its main purpose is still to convert from PyTorch to numpy (even when we are unpacking other data types). If you agree with this definition we should

1. make this explicit in the docs
2. make this explicit in the tests (check for `TypeError` when supplying non-pytorch data, even if nested)

**Comment by BenjaminBossan** [comment]
That sounds very reasonable, I think `to_numpy` shouldn't grow in scope beyond what it delivers at the moment, it was really only meant to be a convenience function.

File: skorch/tests/test_utils.py [hunk 2]
```diff
@@ -174,6 +178,25 @@ def test_tuple(self, to_numpy, x_tuple):
...
+        self.compare_array_to_tensor(x_numpy['b'][1], x_dict['b'][1])
+
+    @pytest.mark.parametrize('x_invalid', [
+        1,
+        [1,2,3],
+        (1,2,3),
+        {'a': 1},
+    ])
+    def test_invalid_inputs(self, to_numpy, x_invalid):
+        " Inputs that are invalid for the scope of to_numpy. "
```
Comment (BenjaminBossan) [inline skorch/tests/test_utils.py:194]: I've never seen this style, neither docstring nor comment. I think a comment would be sufficient here?

Comment (ottonemo) [reply skorch/tests/test_utils.py]: It is still a doc-string but according to PEP257 this style is disregarded. I converted it to a comment.

Example:
```python
>>> def foo():
...     "bla"
...     return 1
... 
>>> foo.__doc__
'bla'
```

Comment (BenjaminBossan) [reply skorch/tests/test_utils.py]: > It is still a doc-string but according to PEP257 this style is disregarded

Interesting, I didn't know, though in hindsight it makes sense.

**Pull request merged**